## DATA CLEANING

In [1]:
import pandas as pd 
# Importing the data - launched phones
df_0 = pd.read_csv("D:\Project-1\Webscraper\proddata.csv")
df_0.head()

battery  capacity  \
0  5000 mah battery with super vooc charging, cha...    5000.0   
1                4500 mah battery with fast charging    4500.0   
2                5000 mah battery with fast charging    5000.0   
3                                   3200 mah battery    3200.0   
4                                   4200 mah battery    4200.0   

                                        connectivity  display  \
0  dual sim: nano + nano (hybrid) with volte support    17.07   
1           dual sim: nano + nano with volte support    16.59   
2           dual sim: nano + nano with volte support    19.81   
3                                  single sim: micro    14.99   
4                                           dual sim    12.70   

  fast_charge_point                                           features  \
0               Yes  side fingerprint sensor |face unlock|3.5 mm he...   
1               Yes  rear fingerprint sensor |face unlock|3.5 mm he...   
2               Yes  side fingerprint sensor |face unlock|no 3.5 mm...   
3                No                      3.5 mm headphone jack present   
4                No  no fingerprint sensor |3.5 mm headphone jack p...   

   front_camera  in_store  launch_date  \
0          16.0     128.0        111.0   
1          20.0     128.0       1389.0   
2          24.0     512.0          NaN   
3           2.1      16.0       3398.0   
4           2.0       4.0       3457.0   

                                            oper_sys  ... prod_brand  \
0                                        android v13  ...    oneplus   
1           android v9.0 (pie) upgradable to v10 (q)  ...     xiaomi   
2                                        android v12  ...    samsung   
3                            android v4.4.2 (kitkat)  ...         lg   
4  android v4.2 (jelly bean) upgradable to v4.4 (...  ...     gionee   

                       prod_name prod_no_of_ratings  prod_price  prod_star  \
0      oneplus nord ce 3 lite 5g              15991       19999        4.1   
1  xiaomi redmi note 8 pro 128gb             230710       15998        4.4   
2     samsung galaxy z fold 6 5g                  0      174990        0.0   
3                     lg g pro 2                244       42021        4.5   
4                  gionee m2 4gb                476       13999        4.0   

                                            prod_url product_availability  \
0  https://pricebaba.com/mobile/oneplus-nord-ce-3...             in stock   
1  https://pricebaba.com/mobile/xiaomi-redmi-note...             in stock   
2  https://pricebaba.com/mobile/samsung-galaxy-z-...             in stock   
3       https://pricebaba.com/mobile/lg-g-pro-2-16gb         out of stock   
4             https://pricebaba.com/mobile/gionee-m2         out of stock   

    ram  rear_camera  refresh_rate  
0   8.0        108.0           120  
1   6.0         64.0            60  
2  16.0         32.0            60  
3   3.0         13.0            60  
4   1.0          8.0            60  

[5 rows x 21 columns]

In [2]:
# Importing the data - Upcoming phones
df_1 = pd.read_csv(r"D:\Project-1\Webscraper\upcomimg.csv")
df_1.head()

battery  capacity  \
0                4500 mah battery with fast charging      4500   
1                                   2200 mah battery      2200   
2                5000 mah battery with fast charging      5000   
3  4700 mah battery with flash charging, charges ...      4700   
4  5000 mah battery with fast charging, charges u...      5000   

                               connectivity  display fast_charge_point  \
0  dual sim: nano + nano with volte support    17.02               Yes   
1                                  dual sim    12.70                No   
2  dual sim: nano + nano with volte support    16.94               Yes   
3  dual sim: nano + nano with volte support    17.22               Yes   
4  dual sim: nano + nano with volte support    17.22               Yes   

                                            features  front_camera  in_store  \
0  side fingerprint sensor |face unlock|no 3.5 mm...           8.0     256.0   
1                      3.5 mm headphone jack present           5.0       8.0   
2  on-screen fingerprint sensor |face unlock|no 3...          16.0     128.0   
3  on-screen fingerprint sensor |face unlock|no 3...          16.0     128.0   
4  side fingerprint sensor |face unlock|3.5 mm he...           8.0     256.0   

   launch_date                                           oper_sys  ...  \
0          NaN                                        android v13  ...   
1          NaN  android v4.2 (jelly bean) upgradable to v4.4 (...  ...   
2          NaN                                        android v13  ...   
3          NaN                                        android v13  ...   
4          NaN                                        android v13  ...   

   prod_brand                   prod_name prod_no_of_ratings  prod_price  \
0       honor                  honor x50i                  0       16690   
1  wickedleak  wickedleak wammy neo youth                  0        8490   
2      xiaomi     xiaomi redmi k60 pro 5g                  0       39290   
3        iqoo              iqoo 11 pro 5g                  0       59190   
4     infinix              infinix hot 30                  0       10999   

   prod_star                                           prod_url  \
0        0.0            https://pricebaba.com/mobile/honor-x50i   
1        0.0  https://pricebaba.com/mobile/wickedleak-wammy-...   
2        0.0  https://pricebaba.com/mobile/xiaomi-redmi-k60-...   
3        0.0        https://pricebaba.com/mobile/iqoo-11-pro-5g   
4        0.0        https://pricebaba.com/mobile/infinix-hot-30   

  product_availability  ram  rear_camera  refresh_rate  
0             in stock  8.0        100.0            90  
1             in stock  1.0         13.0            60  
2             in stock  8.0         54.0           120  
3             in stock  8.0         50.0           144  
4             in stock  8.0         50.0            90  

[5 rows x 21 columns]

In [3]:
# Merging Two dataframes into a single Dataframe (df)
df = pd.concat([df_0, df_1], axis=0, ignore_index=True)

In [4]:
# gathering the basic info 
print(df.columns,end="\n\n")

print(f"No.of mobiles in the scraped data is {df.shape[0]}")


Index(['battery', 'capacity', 'connectivity', 'display', 'fast_charge_point',
       'features', 'front_camera', 'in_store', 'launch_date', 'oper_sys',
       'perform', 'prod_brand', 'prod_name', 'prod_no_of_ratings',
       'prod_price', 'prod_star', 'prod_url', 'product_availability', 'ram',
       'rear_camera', 'refresh_rate'],
      dtype='object')

No.of mobiles in the scraped data is 2718


### Checking for missing values

In [5]:
# analysing missing values
df.isna().sum()

battery                   1
capacity                  5
connectivity              0
display                   0
fast_charge_point         0
features                  0
front_camera              0
in_store                  0
launch_date             628
oper_sys                 12
perform                   3
prod_brand                0
prod_name                 0
prod_no_of_ratings        0
prod_price                0
prod_star                 0
prod_url                  0
product_availability      0
ram                       0
rear_camera               0
refresh_rate              0
dtype: int64

In [6]:
# Since we've upcoming phones , the launched dates will have more missing values 
# because upcoming phones doesn't have launched dates
# instead of removing all missing values --> setting missing launched_date as 0
# --> Then there will be no loss of upcoming phones data
df["launch_date"].fillna(0,inplace=True)

In [7]:
# backuping launch_date column --> for future purpose
launch_date = df["launch_date"]
# now dropping all missing values (na)
df.dropna(axis = 0,inplace=True ,ignore_index = True)

In [8]:
# overview of cleaning missing values and Re-checking
print(df.shape)
df.isna().sum()

(2701, 21)


battery                 0
capacity                0
connectivity            0
display                 0
fast_charge_point       0
features                0
front_camera            0
in_store                0
launch_date             0
oper_sys                0
perform                 0
prod_brand              0
prod_name               0
prod_no_of_ratings      0
prod_price              0
prod_star               0
prod_url                0
product_availability    0
ram                     0
rear_camera             0
refresh_rate            0
dtype: int64

#### No missing values and there are 2701 rows(moblies) of data

In [9]:
# Checking datatypes of all columns
df.dtypes

battery                  object
capacity                float64
connectivity             object
display                 float64
fast_charge_point        object
features                 object
front_camera            float64
in_store                float64
launch_date             float64
oper_sys                 object
perform                  object
prod_brand               object
prod_name                object
prod_no_of_ratings        int64
prod_price                int64
prod_star               float64
prod_url                 object
product_availability     object
ram                     float64
rear_camera             float64
refresh_rate              int64
dtype: object

#### Cleaning data from unreliable information (like sentences,words,...) and assigning scores

In [10]:
# Backuping product_availability
product_availability = df["product_availability"]
# deleting 
del df["battery"]
del df["product_availability"]

In [11]:
#Checking for distinct data in connectivity column
conn = { item for item in df["connectivity"]} # Set comprehension
conn

{' with volte support',
 'dual sim',
 'dual sim + nano (hybrid)',
 'dual sim + nano (hybrid) with volte support',
 'dual sim + nano with volte support',
 'dual sim with dual standby volte support',
 'dual sim with volte support',
 'dual sim: micro',
 'dual sim: micro + micro',
 'dual sim: micro + micro with volte support',
 'dual sim: micro + mini',
 'dual sim: micro + nano',
 'dual sim: micro + nano (hybrid)',
 'dual sim: micro + nano (hybrid) with dual standby volte support',
 'dual sim: micro + nano (hybrid) with volte support',
 'dual sim: micro + nano with volte support',
 'dual sim: micro with volte support',
 'dual sim: mini',
 'dual sim: mini + micro',
 'dual sim: mini + mini',
 'dual sim: mini + nano with volte support',
 'dual sim: nano',
 'dual sim: nano + esim with volte support',
 'dual sim: nano + micro',
 'dual sim: nano + micro hybrid with volte support',
 'dual sim: nano + mini',
 'dual sim: nano + nano',
 'dual sim: nano + nano (hybrid)',
 'dual sim: nano + nano (hybr

In [ ]:
# setting scores for connectivity column
for i in range(df.shape[0]) :
    if "dual" in df["connectivity"][i] :
        df["connectivity"][i] = 1 # dual sim
    else :
        df["connectivity"][i] = 0 # single sim

In [ ]:
# setting scores for fast_charge_point column
for i in range(df.shape[0]) :
    if df["fast_charge_point"][i].lower() == "yes" :
        df["fast_charge_point"][i] = 1 # supports fast charge
    else :
        df["fast_charge_point"][i] = 0 # not supports fast charge   

In [14]:
# Checking for distinct data in oper_sys column
oper = {item for item in df["oper_sys"]} 
oper

{'android',
 'android v10 (q)',
 'android v10 (q) upgradable to v11',
 'android v10 (q) upgradable to v12',
 'android v10 (q) upgradable to v12l',
 'android v10 (q) upgradable to v13',
 'android v11',
 'android v11 upgradable to v12',
 'android v11 upgradable to v13',
 'android v11 upgradable to v14',
 'android v12',
 'android v12 upgradable to v13',
 'android v12 upgradable to v14',
 'android v12l upgradable to v14',
 'android v13',
 'android v13 upgradable to v14',
 'android v14',
 'android v2.0',
 'android v2.1',
 'android v2.2 (froyo)',
 'android v2.3 (gingerbread)',
 'android v2.3 (gingerbread) upgradable to v4.0 (ice cream sandwich)',
 'android v4 (ice cream sandwich)',
 'android v4.0 (ice cream sandwich)',
 'android v4.1 (jelly bean)',
 'android v4.1 (jelly bean) upgradable to v4.2.2 (jelly bean)',
 'android v4.1 (jelly bean) upgradable to v4.3 (jelly bean)',
 'android v4.1.1 (jelly bean)',
 'android v4.1.1 (jelly bean) upgradable to v4.2.2 (jelly bean)',
 'android v4.1.2 (jelly

#### Formatting data into simpler form 

In [15]:
oper_list = { item.split()[0] for item in oper}
oper_list.remove("windows") # only windows version is on the index [2]
oper_list 

{'android',
 'blackberry',
 'emui',
 'flyme',
 'harmonyos',
 'ios',
 'kai',
 'nokia',
 'symbian'}

In [ ]:
# Gathering oper_sys(name + version) format
for i in range(df.shape[0]) :
    if 'windows' in df["oper_sys"][i] :
        try :
            df["oper_sys"][i] = "windows " + df["oper_sys"][i].split()[2].split(".")[0].replace("v","")
        except IndexError :
            df["oper_sys"][i] = "windows"    # (i.e) windows 10
    else :        
        for term in oper_list :
            if term in df["oper_sys"][i] :
                try :
                    df["oper_sys"][i] = term + " " + df["oper_sys"][i].split()[1].split(".")[0].replace("v","")
                except IndexError :
                    df["oper_sys"][i] = term  # (i.e) android 11
        

In [17]:
# checking new formatted data 
oper_sort = { term for term in df["oper_sys"]}
oper_sort

{'android',
 'android 10',
 'android 11',
 'android 12',
 'android 12l',
 'android 13',
 'android 14',
 'android 2',
 'android 4',
 'android 5',
 'android 6',
 'android 7',
 'android 8',
 'android 9',
 'blackberry',
 'blackberry 10',
 'blackberry 5',
 'blackberry 6',
 'blackberry 7',
 'emui 13',
 'flyme os',
 'harmonyos',
 'ios 10',
 'ios 11',
 'ios 12',
 'ios 13',
 'ios 14',
 'ios 15',
 'ios 16',
 'ios 4',
 'ios 7',
 'ios 8',
 'ios 9',
 'kai os',
 'nokia asha',
 'symbian',
 'symbian belle',
 'windows 10',
 'windows 7',
 'windows 8',
 'windows upgradable'}

##### Altering some data for our convenient

In [18]:
# android 12l --> android 12
import numpy as np
b = np.where(df["oper_sys"] == "android 12l")
b

(array([ 376,  786, 1959], dtype=int64),)

In [ ]:
for i in [376,  786, 1959] :
     df["oper_sys"][i] = df["oper_sys"][i].replace("l","")

In [20]:
# windows upgradable --> windows
np.where(df["oper_sys"].str.contains("windows upgradable"))

(array([280], dtype=int64),)

In [ ]:
df["oper_sys"][280] = "windows 8"

In [22]:
# symbian belle --> symbian
np.where(df["oper_sys"].str.contains("symbian belle"))

(array([1625], dtype=int64),)

In [ ]:
df["oper_sys"][1625] = "symbian"

In [24]:
oper_sort_clean = {term for term in df["oper_sys"] }
# writing the distinct oper_sys to .csv file
# with open("D:\Project-1\Webscraper\oper_sys.csv","w") as oper_csv :
#     for item in oper_sort_clean :
#         oper_csv.write(item + "," + "\n")
#     oper_csv.close()    


In [25]:
# setting scores for each oper_sys with the help of web (manually) 
oper_data_frame = pd.read_csv("D:\Project-1\Webscraper\oper_sys.csv",header=None) # .csv --> dataframe
oper_data_frame.sort_values(by=1,ascending=False,inplace=True,ignore_index=True) 
oper_data_frame


0   1
0          ios 16  36
1          ios 15  35
2      android 14  34
3      android 12  33
4      android 13  33
5       harmonyos  32
6          kai os  31
7      windows 10  30
8   blackberry 10  29
9    blackberry 7  28
10       flyme os  27
11         ios 14  26
12     android 11  25
13        emui 13  25
14         ios 13  24
15     android 10  23
16      android 9  22
17         ios 12  21
18      android 8  20
19      android 7  19
20         ios 11  18
21         ios 10  17
22      android 6  16
23          ios 9  15
24      android 5  14
25          ios 8  13
26      android 4  12
27          ios 7  11
28      windows 8  10
29   blackberry 6   9
30   blackberry 5   8
31      windows 7   7
32        symbian   6
33          ios 4   5
34      android 2   4
35     nokia asha   3
36        android   2
37     blackberry   1

In [ ]:
# setting scores with oper_data_frame
for i in range(df.shape[0]) :
    for j in range(oper_data_frame.shape[0]) :
        if df["oper_sys"][i] == oper_data_frame[0][j] :
            df["oper_sys"][i] = oper_data_frame[1][j]

In [27]:
# checking distinct perform data
perform = { term for term in df["perform"]}
perform

{'128 mb ram',
 '200 mb internal storage, expandable upto 32 gb',
 '256 mb internal storage, expandable upto 32 gb',
 '256 mb ram',
 '512 mb ram',
 'apple a10 fusion apl1024 quad core processor',
 'apple a11 bionic hexa core processor',
 'apple a12 bionic hexa core processor',
 'apple a13 bionic hexa core processor',
 'apple a14 bionic hexa core processor',
 'apple a15 bionic hexa core processor',
 'apple a16 bionic hexa core processor',
 'apple a4 apl0398 single core processor',
 'apple a6 apl0589 dual core processor',
 'apple a7 apl0698 dual core processor',
 'apple a8 apl1011 dual core processor',
 'apple a8 dual core processor',
 'apple a9 apl0898 dual core processor',
 'apple a9 apl1022 dual core processor',
 'broadcom bcm21664t dual core processor',
 'broadcom bcm23550 quad core processor',
 'broadcom bcm28155 dual core processor',
 'dual core processor',
 'google tensor g2 octa core processor',
 'google tensor g3 octa core processor',
 'google tensor octa core processor',
 'hisi

In [28]:
# converting into simpler form
perform = {term.split()[0]  for term in df["perform"]}
perform

{'128',
 '200',
 '256',
 '512',
 'apple',
 'broadcom',
 'dual',
 'google',
 'hisilicon',
 'huawei',
 'intel',
 'marvell',
 'mediatek',
 'nvidia',
 'octa',
 'quad',
 'qualcomm',
 'samsung',
 'single',
 'spreadtrum',
 'st-ericsson',
 'ti',
 'tru-octa',
 'unisoc'}

In [29]:
# writing distinct simpler perform data as .csv file
# with open("D:\Project-1\Webscraper\perform.csv","w") as perform_text :
#     for term in perform :
#         perform_text.write(term + "," + "\n")
#     perform_text.close()    

In [30]:
# assigning scores with the help of web
perform_data_frame = pd.read_csv("D:\Project-1\Webscraper\perform.csv",header=None)
perform_data_frame.sort_values(by = 1,ignore_index=True,inplace=True,ascending=False) # .csv --> dataframe
perform_data_frame

0   1
0         apple  15
1      qualcomm  14
2       samsung  13
3      mediatek  12
4         intel  11
5        nvidia  10
6     hisilicon   9
7      broadcom   8
8        google   7
9        unisoc   6
10      marvell   5
11  st-ericsson   4
12       huawei   3
13   spreadtrum   2

In [ ]:
# setting scores with perform_data_frame
for i in range(df.shape[0]) :  
    if str(df["perform"][i]).split()[0] not in [ term for term in perform_data_frame[0] ]:
        df["perform"][i] = 1.0
    else :
        for j in range(perform_data_frame.shape[0]) :
            if str(df["perform"][i]).split()[0] == perform_data_frame[0][j] :
                df["perform"][i] = perform_data_frame[1][j]    
    

##### Again deleting unwanted data

In [32]:
del df["features"]
del df["prod_url"]
del df["prod_no_of_ratings"]

In [33]:
# attaching back-uped data
df["launch_date"] = launch_date
df["product_availability"] = product_availability

In [34]:
# (df) dataframe --> .csv file cleaned data 
df.to_csv("D:\Project-1\Webscraper\proddata_R.csv",index=None)

### DATACLEANING -  finished

#### After Clustering(grouping) the cleaned data based on Agglomerative hierarchical clustering-Euclidean distance-Ward method-ANOVA. The Cluster assignment is done and saved on a new .csv file as cluster.csv

In [35]:

# Creating dataframe
import pandas as pd
new_df = pd.read_csv("D:\Project-1\Webscraper\cluster.csv")

In [36]:
new_df

capacity  display  fast_charge_point  front_camera  in_store  oper_sys  \
0         5000    17.07                  1          16.0     128.0        33   
1         4500    16.59                  1          20.0     128.0        22   
2         5000    19.81                  1          24.0     512.0        33   
3         3200    14.99                  0           2.1      16.0        12   
4         4200    12.70                  0           2.0       4.0        12   
...        ...      ...                ...           ...       ...       ...   
2696      5000    17.22                  1          16.0     128.0        33   
2697      5000    16.33                  1          16.0     128.0        33   
2698      4800    17.12                  1          32.0     256.0        33   
2699      5000    17.25                  1           8.0     128.0        33   
2700      5100    16.94                  1          60.0     256.0        33   

      perform  prod_price   ram  rear_camera  refresh_rate  cluster  \
0          14       19999   8.0        108.0           120        1   
1          12       15998   6.0         64.0            60        2   
2          14      174990  16.0         32.0            60        3   
3          14       42021   3.0         13.0            60        4   
4          12       13999   1.0          8.0            60        5   
...       ...         ...   ...          ...           ...      ...   
2696       14       31840   8.0         64.0           120        7   
2697       12       15999   6.0        108.0            90        2   
2698       12       54190  12.0         50.0           120        8   
2699       12       12290   4.0         50.0            90        5   
2700       14       65490   8.0         50.0           165        8   

                          prod_name prod_brand  prod_star  launch_date  \
0         oneplus nord ce 3 lite 5g    oneplus        4.1          111   
1     xiaomi redmi note 8 pro 128gb     xiaomi        4.4         1389   
2        samsung galaxy z fold 6 5g    samsung        0.0            0   
3                        lg g pro 2         lg        4.5         3398   
4                     gionee m2 4gb     gionee        4.0         3457   
...                             ...        ...        ...          ...   
2696                  vivo v29 lite       vivo        0.0            0   
2697          xiaomi redmi note 12s     xiaomi        0.0            0   
2698                   oppo find x6       oppo        0.0            0   
2699                xiaomi redmi 12     xiaomi        0.0            0   
2700        motorola edge plus 2023   motorola        0.0            0   

     product_availability  
0                in stock  
1                in stock  
2                in stock  
3            out of stock  
4            out of stock  
...                   ...  
2696             in stock  
2697             in stock  
2698             in stock  
2699             in stock  
2700             in stock  

[2701 rows x 17 columns]

In [37]:
# Dropping unrated phones from each cluster [ phones have 0.0 stars]
import numpy as np
new_df.drop(np.where(new_df["prod_star"] == 0.0)[0],axis = 0,inplace = True)

In [38]:
# Re-Checking
(new_df["prod_star"] == 0.0).sum()

0

In [39]:
# calculating remaining phone's ( mean,median,mode and variance of prod_star ) for each cluster
stars_df = new_df.groupby("cluster")["prod_star"].aggregate(["var","mean","median",lambda x: x.mode()[0]])

In [40]:
# for future purpose --> adding cluster column
stars_df["cluster"] = [1,2,3,4,5,6,7,8,9,10]
stars_df

var      mean  median  <lambda_0>  cluster
cluster                                                 
1        0.243803  4.082530     4.2         4.3        1
2        0.199171  4.103433     4.2         4.3        2
3        0.159547  4.470833     4.6         4.6        3
4        0.530320  4.009524     4.2         4.2        4
5        0.191432  4.037361     4.2         4.3        5
6        0.361493  4.142169     4.3         4.4        6
7        0.377169  4.100625     4.3         4.3        7
8        0.219485  4.258667     4.3         4.4        8
9        0.176837  4.370000     4.5         4.5        9
10       0.190058  4.468421     4.6         4.5       10

In [41]:

# Creating dataframe
import pandas as pd
final_df = pd.read_csv("D:\Project-1\Webscraper\cluster.csv")
final_df

capacity  display  fast_charge_point  front_camera  in_store  oper_sys  \
0         5000    17.07                  1          16.0     128.0        33   
1         4500    16.59                  1          20.0     128.0        22   
2         5000    19.81                  1          24.0     512.0        33   
3         3200    14.99                  0           2.1      16.0        12   
4         4200    12.70                  0           2.0       4.0        12   
...        ...      ...                ...           ...       ...       ...   
2696      5000    17.22                  1          16.0     128.0        33   
2697      5000    16.33                  1          16.0     128.0        33   
2698      4800    17.12                  1          32.0     256.0        33   
2699      5000    17.25                  1           8.0     128.0        33   
2700      5100    16.94                  1          60.0     256.0        33   

      perform  prod_price   ram  rear_camera  refresh_rate  cluster  \
0          14       19999   8.0        108.0           120        1   
1          12       15998   6.0         64.0            60        2   
2          14      174990  16.0         32.0            60        3   
3          14       42021   3.0         13.0            60        4   
4          12       13999   1.0          8.0            60        5   
...       ...         ...   ...          ...           ...      ...   
2696       14       31840   8.0         64.0           120        7   
2697       12       15999   6.0        108.0            90        2   
2698       12       54190  12.0         50.0           120        8   
2699       12       12290   4.0         50.0            90        5   
2700       14       65490   8.0         50.0           165        8   

                          prod_name prod_brand  prod_star  launch_date  \
0         oneplus nord ce 3 lite 5g    oneplus        4.1          111   
1     xiaomi redmi note 8 pro 128gb     xiaomi        4.4         1389   
2        samsung galaxy z fold 6 5g    samsung        0.0            0   
3                        lg g pro 2         lg        4.5         3398   
4                     gionee m2 4gb     gionee        4.0         3457   
...                             ...        ...        ...          ...   
2696                  vivo v29 lite       vivo        0.0            0   
2697          xiaomi redmi note 12s     xiaomi        0.0            0   
2698                   oppo find x6       oppo        0.0            0   
2699                xiaomi redmi 12     xiaomi        0.0            0   
2700        motorola edge plus 2023   motorola        0.0            0   

     product_availability  
0                in stock  
1                in stock  
2                in stock  
3            out of stock  
4            out of stock  
...                   ...  
2696             in stock  
2697             in stock  
2698             in stock  
2699             in stock  
2700             in stock  

[2701 rows x 17 columns]

In [ ]:
# Assigning median of prod_star of a cluster for 0.0 starred phones in the same cluster
for i in range(final_df.shape[0]) :
    if final_df["prod_star"][i] == 0.0 :
        for j in range(1,11) :
            if final_df["cluster"][i] == stars_df["cluster"][j] :
                final_df["prod_star"][i] = stars_df["median"][j]
        

In [43]:
# Final data which contains prod_stars for all mobile phones
final_df

capacity  display  fast_charge_point  front_camera  in_store  oper_sys  \
0         5000    17.07                  1          16.0     128.0        33   
1         4500    16.59                  1          20.0     128.0        22   
2         5000    19.81                  1          24.0     512.0        33   
3         3200    14.99                  0           2.1      16.0        12   
4         4200    12.70                  0           2.0       4.0        12   
...        ...      ...                ...           ...       ...       ...   
2696      5000    17.22                  1          16.0     128.0        33   
2697      5000    16.33                  1          16.0     128.0        33   
2698      4800    17.12                  1          32.0     256.0        33   
2699      5000    17.25                  1           8.0     128.0        33   
2700      5100    16.94                  1          60.0     256.0        33   

      perform  prod_price   ram  rear_camera  refresh_rate  cluster  \
0          14       19999   8.0        108.0           120        1   
1          12       15998   6.0         64.0            60        2   
2          14      174990  16.0         32.0            60        3   
3          14       42021   3.0         13.0            60        4   
4          12       13999   1.0          8.0            60        5   
...       ...         ...   ...          ...           ...      ...   
2696       14       31840   8.0         64.0           120        7   
2697       12       15999   6.0        108.0            90        2   
2698       12       54190  12.0         50.0           120        8   
2699       12       12290   4.0         50.0            90        5   
2700       14       65490   8.0         50.0           165        8   

                          prod_name prod_brand  prod_star  launch_date  \
0         oneplus nord ce 3 lite 5g    oneplus        4.1          111   
1     xiaomi redmi note 8 pro 128gb     xiaomi        4.4         1389   
2        samsung galaxy z fold 6 5g    samsung        4.6            0   
3                        lg g pro 2         lg        4.5         3398   
4                     gionee m2 4gb     gionee        4.0         3457   
...                             ...        ...        ...          ...   
2696                  vivo v29 lite       vivo        4.3            0   
2697          xiaomi redmi note 12s     xiaomi        4.2            0   
2698                   oppo find x6       oppo        4.3            0   
2699                xiaomi redmi 12     xiaomi        4.2            0   
2700        motorola edge plus 2023   motorola        4.3            0   

     product_availability  
0                in stock  
1                in stock  
2                in stock  
3            out of stock  
4            out of stock  
...                   ...  
2696             in stock  
2697             in stock  
2698             in stock  
2699             in stock  
2700             in stock  

[2701 rows x 17 columns]

In [44]:
# Re-Checking
(final_df["prod_star"] == 0.0).sum()

0

### INFERENCE PART
#### Our aim is to list the brands based on their products performance
#### And also listing top brands based on their no of products launched ( Reputed Brands )

In [45]:
# Analysing each brands for their no.of products
brand_name_set = { term for term in final_df["prod_brand"] }
brand_dict = list()
for brand in brand_name_set :
    brand_dict.append({ "brand" : brand , "no_of_prod" : len(np.where(final_df["prod_brand"] == brand)[0])})
    


In [46]:
# brands and their products as dictionary
brand_dict

[{'brand': 'sony', 'no_of_prod': 49},
 {'brand': 'forme', 'no_of_prod': 1},
 {'brand': 'itel', 'no_of_prod': 1},
 {'brand': 'infinix', 'no_of_prod': 64},
 {'brand': 'blackberry', 'no_of_prod': 21},
 {'brand': 'meizu', 'no_of_prod': 8},
 {'brand': 'asus', 'no_of_prod': 60},
 {'brand': 'zopo', 'no_of_prod': 1},
 {'brand': 'zen', 'no_of_prod': 1},
 {'brand': 'lenovo', 'no_of_prod': 27},
 {'brand': 'gionee', 'no_of_prod': 24},
 {'brand': 'motorola', 'no_of_prod': 129},
 {'brand': 'lg', 'no_of_prod': 44},
 {'brand': 'blu', 'no_of_prod': 1},
 {'brand': 'videocon', 'no_of_prod': 1},
 {'brand': 'poco', 'no_of_prod': 64},
 {'brand': 'reliance', 'no_of_prod': 1},
 {'brand': 'idea', 'no_of_prod': 1},
 {'brand': 'nubia', 'no_of_prod': 12},
 {'brand': 'coolpad', 'no_of_prod': 6},
 {'brand': 'leeco', 'no_of_prod': 2},
 {'brand': 'meiigoo', 'no_of_prod': 2},
 {'brand': 'tecno', 'no_of_prod': 73},
 {'brand': 'yuho', 'no_of_prod': 2},
 {'brand': 'elephone', 'no_of_prod': 6},
 {'brand': 'nokia', 'no_of_

#### Taking Top 10 brands based on the maximum no of launches of their products

In [47]:
brand_df = pd.DataFrame(brand_dict).sort_values("no_of_prod",ascending=False)[0:10]

In [48]:
brand_df

brand  no_of_prod
73   samsung         347
47      vivo         237
76    realme         226
70    xiaomi         219
50      oppo         173
11  motorola         129
28     apple         112
64   oneplus         101
25     nokia          76
80      iqoo          74

In [49]:
# collecting those brands
results_list = list()
for brand in brand_df["brand"] :
        results_list.append(brand)

In [50]:
# Back up starred data as .csv
final_df.to_csv("D:\Project-1\Webscraper\\final.csv",index= None)


note : launch_date is the no.of days that the product is since launched to 1st july 2023

In [51]:
# Removing upcoming phones --> upcoming phones have launched_date 0
final_df.drop(np.where(final_df["launch_date"] == 0)[0],axis = 0,inplace = True)

In [52]:
# grouping based prod_brands and calculating (mean,median,mode and variance) of star-rating for each brand (all brands)
results_df = final_df.groupby("prod_brand")["prod_star"].aggregate(["var","mean","median",lambda x: x.mode()[0]]) 

In [53]:
results_df

var      mean  median  <lambda_0>
prod_brand                                        
10.or       0.845000  4.350000    4.35         3.7
agm         0.202857  4.242857    4.20         4.2
alcatel     0.083333  4.033333    4.20         4.2
apple       0.059688  4.497000    4.55         4.6
asus        0.074072  4.134615    4.10         4.3
...              ...       ...     ...         ...
yota        0.320000  4.600000    4.60         4.2
yu          0.023333  3.833333    3.80         3.7
yuho        0.720000  3.600000    3.60         3.0
zen              NaN  4.700000    4.70         4.7
zopo             NaN  3.600000    3.60         3.6

[81 rows x 4 columns]

In [54]:
# list of Top 10 brands
final_result_df = results_df.loc[results_list].sort_values("mean",ascending=False)

In [55]:
final_result_df # based on mean

var      mean  median  <lambda_0>
prod_brand                                        
apple       0.059688  4.497000    4.55         4.6
realme      0.082935  4.324044    4.40         4.4
oneplus     0.076113  4.275676    4.30         4.3
oppo        0.096978  4.255469    4.30         4.4
vivo        0.113808  4.241765    4.30         4.3
iqoo        0.024533  4.236170    4.20         4.2
xiaomi      0.031692  4.222603    4.20         4.2
samsung     0.161949  4.106620    4.20         4.3
motorola    0.154713  4.079747    4.20         4.2
nokia       0.191642  3.907407    4.00         4.2